scrapy/core/scheduler.py

```diff
@@ -3,14 +3,26 @@
 import logging
 from os.path import join, exists
 
-from scrapy.utils.reqser import request_to_dict, request_from_dict
-from scrapy.utils.misc import load_object
+from scrapy.utils.httpobj import urlparse_cached
 from scrapy.utils.job import job_dir
+from scrapy.utils.misc import load_object
+from scrapy.utils.reqser import request_to_dict, request_from_dict
+from queuelib import RoundRobinQueue
 
 logger = logging.getLogger(__name__)
 
+def _make_file_safe(string):
+    """
+    Make key file safe.
+
+    Warning: not a general function.  Works for all numbers and the vast majority of domains.
+    Collisions in file names will only degrade (midly) performance,
+    not lead to incorrect results.
```

Comment:
I'm still not convinced we should use a function with known limitations, while there is an easy workaround - e.g. add md5 or sha1 hash (or its part?) of the original content to the end. I'm not sure, but replacing a list comprehension with a simpler regex + an md5 hash may even make this code faster. 

We also probably want to limit string length, as file systems can have problems with that.

If we're to be strict, `isalnum` check can also be problematic, as it doesn't guarantee ascii.

Anyways, I agree that the current code works in most cases, and probably it is not a big deal to leave it as-is, but it doesn't seem too hard to make it work in all cases: be more aggressive with the human-readable part, add a hash.

Reply:
Also, a nitpick: one more blank line is needed before this function according to pep8.